### DISCUSSION TITLE
Easiest way to draw a thread

### DISCUSSION THREAD

**Discussion by Jamesgt** [sgenoud/replicad#149] [How To]
It is hard to understand the way how a thread is drawn in this example https://replicad.xyz/docs/examples/occt-bottle.

Is there an easy solution for a simple thread?
You know, I just need to revolve and lift up in the same time the triangle.
Probably sweep is what I need to learn, but was not able to till now. :(

In this example you can see the path:

```js
/** @type { import("replicad") } */ const R = replicad

const fuseAll = (...shapes) => shapes.reduce((acc, shape) => shape ? (acc ? acc.fuse(shape) : shape) : acc, shapes[0]);
const stack = (elements) => {
  let offset = 0
  return elements.map((element) => {
    const { depth: z } = element.shape.boundingBox
    element.shape = element.shape.translate([0, 0, offset])
    offset += z
    return element
  })
}

const main = () => {
  const base = R.makeCylinder(58, 2.85)
    .cut(R.makeCylinder(58 - 1.5, 1.6))
    .cut(R.makeCylinder(44.65 - 1.6, 10))
  const thread = R.drawPolysides(0.8, 3).translate(44.65 - 1.6)
    .sketchOnPlane('XY')
    .extrude(7, { twistAngle: 3 * 360 })
  const top = R.makeCylinder(44.65, 7)
    .cut(R.makeCylinder(44.65 - 1.6, 10))
  //.fuse(thread)

  return stack([
    { name: 'base', shape: base, color: 'grey' },
    { name: 'top', shape: top, color: 'cyan' },
    { name: 'thread', shape: thread, color: 'darkslategrey' },
  ])
}
```

**Comment by sgenoud** [comment]
You were right you can use an helix to sweep a profile (in that case a triangle, but you can be more creative, for instance with 45 degrees angles for better 3D printing):

```js
const {sketchHelix, draw, Plane} = replicad

function simpleThread({ innerRadius, pitch, threadHeight, toothSide }) {
  const drawProfile = (plane, origin) => {
    const originPlane = new Plane(origin, [0, 0, -1], plane.zDir);
    return draw()
      .hLine(-toothSide)
      .polarLine(toothSide, 60)
      .close()
      .sketchOnPlane(originPlane);
  };

  const threadBase = sketchHelix(
    pitch,
    threadHeight,
    innerRadius,
    [0, 0, 0],
    [0, 0, 1],
    false
  );
  return threadBase.sweepSketch(drawProfile, { frenet: true });
}
```

I then fuse it this was with the main cylinder:


```js
const thread = simpleThread({
  innerRadius: 44.65,
  pitch: 1.7,
  threadHeight: 6.2,
  toothSide: 0.8,
});

const top = R.makeCylinder(44.65, 7)
  .cut(R.makeCylinder(44.65 - 1.6, 10))
  .rotate(2)
  .fuse(thread.clone(), { optimisation: "sameFace" });
```

Note that I rotate the cylinder of 2 degrees to avoid having the end of our thread start on the cylinder seam which breaks the kernel. I also use the `"sameFace"` optimisation to make the fusion work.

This is the easiest way to do this - but I am not sure this is the best. [The freeCAD wiki](https://wiki.freecad.org/Thread_for_Screw_Tutorial/en) has some different method and tricks that translate quite well on what can be done in replicad.

Ideally, one could port [this code](https://github.com/gumyr/cq_warehouse/blob/main/src/cq_warehouse/thread.py) to replicad to make it easy for anyone to import helper functions as a library.
